File: engines/go/csv.go
package main

import (
	"bufio"
	"encoding/csv"
	"fmt"
	"os"
	"strings"
)

type CSVData struct {
	Headers []string
	Rows    [][]string
}

func readCSV(path string) (*CSVData, error) {
	f, err := os.Open(path)
	if err != nil {
		return nil, fmt.Errorf("open: %w", err)
	}
	defer f.Close()

	var lines []string
	scanner := bufio.NewScanner(f)
	scanner.Buffer(make([]byte, 1024*1024), 1024*1024)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}
	if err := scanner.Err(); err != nil {
		return nil, fmt.Errorf("scan: %w", err)
	}
	if len(lines) == 0 {
		return nil, fmt.Errorf("csv is empty")
	}

	headers, err := parseWrappedLine(lines[0])
	if err != nil {
		return nil, fmt.Errorf("parse headers: %w", err)
	}
	if len(headers) > 0 {
		headers[0] = strings.TrimPrefix(headers[0], "\xef\xbb\xbf")
	}

	var rows [][]string
	for i, line := range lines[1:] {
		if strings.TrimSpace(line) == "" {
			continue
		}
		row, err := parseWrappedLine(line)
		if err != nil {
			return nil, fmt.Errorf("parse row %d: %w", i+1, err)
		}
		rows = append(rows, row)
	}

	return &CSVData{Headers: headers, Rows: rows}, nil
}

func parseWrappedLine(line string) ([]string, error) {
	unwrapped  := unwrapOuter(line)
	normalized := strings.ReplaceAll(unwrapped, `""`, `"`)

	reader                  := csv.NewReader(strings.NewReader(normalized))
	reader.LazyQuotes        = true
	reader.TrimLeadingSpace  = true

	fields, err := reader.Read()
	if err != nil {
		return nil, fmt.Errorf("csv parse %q: %w", normalized, err)
	}

	for i, f := range fields {
		fields[i] = cleanStr(f)
	}

	return fields, nil
}

func unwrapOuter(s string) string {
	s = strings.TrimSpace(s)
	if len(s) >= 2 && s[0] == '"' && s[len(s)-1] == '"' {
		return s[1 : len(s)-1]
	}
	return s
}

func cleanStr(s string) string {
	for {
		stripped := strings.TrimSpace(s)
		stripped  = strings.Trim(stripped, `"`)
		stripped  = strings.Trim(stripped, "`")
		stripped  = strings.Trim(stripped, `'`)
		stripped  = strings.TrimSpace(stripped)
		if stripped == s {
			break
		}
		s = stripped
	}
	return s
}

File: engines/rust/src/csv.rs
#[derive(Debug)]
pub struct CSVData {
    pub headers: Vec<String>,
    pub rows:    Vec<Vec<String>>,
}

// Снимает все слои кавычек — аналог cleanStr в Go
pub fn clean_identifier(s: &str) -> String {
    let mut result = s.to_string();
    loop {
        let prev = result.clone();
        let stripped = result.trim();
        let stripped = stripped
            .trim_matches('"')
            .trim_matches('`')
            .trim_matches('\'')
            .trim();
        result = stripped.to_string();
        if result == prev {
            break;
        }
    }
    result
}

// Снимает внешние кавычки
fn unwrap_outer(line: &str) -> String {
    let s = line.trim();
    if s.starts_with('"') && s.ends_with('"') && s.len() >= 2 {
        s[1..s.len() - 1].to_string()
    } else {
        s.to_string()
    }
}

// Заменяет "" на "
fn replace_double_quotes(s: &str) -> String {
    s.replace("\"\"", "\"")
}

// Стандартный CSV парсер с поддержкой кавычек
fn parse_csv_line(line: &str) -> Vec<String> {
    let mut fields    = Vec::new();
    let mut current   = String::new();
    let mut in_quotes = false;
    let chars: Vec<char> = line.chars().collect();
    let mut i = 0;

    while i <= chars.len() {
        let c = chars.get(i).copied();
        match c {
            Some('"') => {
                if in_quotes && chars.get(i + 1) == Some(&'"') {
                    current.push('"');
                    i += 1;
                } else {
                    in_quotes = !in_quotes;
                }
            }
            Some(',') if !in_quotes => {
                fields.push(current.clone());
                current.clear();
            }
            None => {
                fields.push(current.clone());
                current.clear();
            }
            Some(ch) => current.push(ch),
        }
        i += 1;
    }

    fields
}

// Парсит строку в формате Wireshark CSV
fn parse_wrapped_line(line: &str) -> Vec<String> {
    let unwrapped  = unwrap_outer(line);
    let normalized = replace_double_quotes(&unwrapped);
    parse_csv_line(&normalized)
        .into_iter()
        .map(|f| clean_identifier(&f))
        .collect()
}

pub fn csv_read(path: &str) -> anyhow::Result<CSVData> {
    let content = std::fs::read_to_string(path)
        .map_err(|e| anyhow::anyhow!("csv_read: {}", e))?;

    let mut lines: Vec<&str> = content
        .lines()
        .filter(|l| !l.trim().is_empty())
        .collect();

    if lines.is_empty() {
        anyhow::bail!("csv_read: file is empty");
    }

    let mut headers = parse_wrapped_line(lines[0]);

    // Убираем BOM
    if let Some(first) = headers.first_mut() {
        *first = first.trim_start_matches('\u{FEFF}').to_string();
    }

    lines.remove(0);
    let rows = lines.iter().map(|l| parse_wrapped_line(l)).collect();

    Ok(CSVData { headers, rows })
}